PZNS_Framework/media/lua/client/03_mod_core/PZNS_NPCFaction.lua

```diff
@@ -0,0 +1,90 @@
+require("00_references/init")
+
+---@class NPCFaction
+---@field factionID factionID Unique faction identifier
+---@field name string Faction name
+---@field leaderID survivorID NPC ID of faction leader
+---@field memberCount integer number of members in faction
+---@field members table<survivorID> list of faction members
+local NPCFaction = {}
+
+
+---Creates new faction with provided parameters and puts it to registry
+---@param factionID factionID
+---@param name string
+---@param leaderID? survivorID
+---@param members? table<survivorID>
+---@return table faction Created faction
+function NPCFaction:new(
+    factionID,
+    name,
+    leaderID,
+    members
+)
+    local faction = {}
+    setmetatable(faction, self)
+    self.__index = self
+
+    faction.factionID = factionID
+    faction.name = name
+    faction.leaderID = leaderID
+    faction.members = members
+    faction.memberCount = #faction.members
+    return faction
+end
+
+---Sets new leader for faction
+---@param newLeaderID factionID
+function NPCFaction:setLeader(newLeaderID)
+    self.leaderID = newLeaderID
+end
+
+---Get list of faction members
+---@return table<survivorID>
+function NPCFaction:getMembers()
+    local res = {}
+    for i = 1, #self.members do
+        res[i] = self.members[i]
+    end
+    return res
+end
+
+---Get number of group members
+function NPCFaction:getMemberCount()
+    return self.memberCount
+end
+
+---Added memberID to faction members
+---@param memberID survivorID NPC survivorID of newly added member
+function NPCFaction:addMember(memberID)
+    if not memberID then return end
+    self.members[self.memberCount + 1] = memberID
```

Comment:
This bothers me a bit, but please correct me if I am wrong.
Let's say player group has 5 members.

PlayerGroup[1] = "player0"
PlayerGroup[2] = "PZNS_Member_2"
PlayerGroup[3] = "PZNS_Member_3"
PlayerGroup[4] = "PZNS_Member_4"
PlayerGroup[5] = "PZNS_Member_5"

Now assume we removed "PZNS_Member_2".
The memberCount would be down to 4.

PlayerGroup[1] = "player0"
PlayerGroup[3] = "PZNS_Member_3"
PlayerGroup[4] = "PZNS_Member_4"
PlayerGroup[5] = "PZNS_Member_5"

If we add a new member,  "PZNS_Member_6", the code logic would look like this?
PlayerGroup[1] = "player0"
PlayerGroup[3] = "PZNS_Member_3"
PlayerGroup[4] = "PZNS_Member_4"
**PlayerGroup[5] = "PZNS_Member_5"**
**PlayerGroup[5] = "PZNS_Member_6"**

Reply:
> This bothers me a bit, but please correct me if I am wrong. Let's say player group has 5 members.
> 
> PlayerGroup[1] = "player0" PlayerGroup[2] = "PZNS_Member_2" PlayerGroup[3] = "PZNS_Member_3" PlayerGroup[4] = "PZNS_Member_4" PlayerGroup[5] = "PZNS_Member_5"
> 
> Now assume we removed "PZNS_Member_2". The memberCount would be down to 4.
> 
> PlayerGroup[1] = "player0" PlayerGroup[3] = "PZNS_Member_3" PlayerGroup[4] = "PZNS_Member_4" PlayerGroup[5] = "PZNS_Member_5"
> 
> If we add a new member, "PZNS_Member_6", the code output would look like this? PlayerGroup[1] = "player0" PlayerGroup[3] = "PZNS_Member_3" PlayerGroup[4] = "PZNS_Member_4" **PlayerGroup[5] = "PZNS_Member_5"** **PlayerGroup[5] = "PZNS_Member_6"**
> 
> Just trying to understand how the index would work in cases where memberCount can change based on in-game events.

removeMember uses table.remove, which reorders the table elements
using the example group:
```
PlayerGroup[1] = "player0"
PlayerGroup[2] = "PZNS_Member_2"
PlayerGroup[3] = "PZNS_Member_3"
PlayerGroup[4] = "PZNS_Member_4"
PlayerGroup[5] = "PZNS_Member_5"
```
after removing PZNS_Member_2 with table.remove it becomes:
```
PlayerGroup[1] = "player0"
PlayerGroup[2] = "PZNS_Member_3"
PlayerGroup[3] = "PZNS_Member_4"
PlayerGroup[4] = "PZNS_Member_5"
```
so memberCount will still point to the last slot